lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -123,19 +125,20 @@ def relationship_value_for(serializer, options = {})
         end
 
         def relationships_for(serializer)
-          Hash[serializer.associations.map { |association| [association.key, { data: relationship_value_for(association.serializer, association.options) }] }]
+          serializer.associations.each_with_object({}) do |association, hash|
+            hash[association.key] = { data: relationship_value_for(association.serializer, association.options) }
+          end
         end
 
-        def included_for(serializer)
-          included.flat_map { |inc|
-            association = serializer.associations.find { |assoc| assoc.key == inc.first }
-            _included_for(association.serializer, inc.second) if association
+        def included_for
+          serializer.associations(@include_tree).flat_map { |association|
```

Comment:
This is where the magic happens. `Serializer#associations` now takes an `IncludeTree` as an optional parameter in order to filter out unwanted associations.